tedge/src/cli/connect/command.rs

```diff
@@ -224,51 +223,66 @@ async fn check_device_status_c8y(port: u16) -> Result<DeviceStatus, ConnectError
     return Ok(DeviceStatus::MightBeNew);
 }
 
-async fn create_device(port: u16) -> Result<DeviceStatus, ConnectError> {
+fn create_device(port: u16) -> Result<DeviceStatus, ConnectError> {
     const C8Y_TOPIC_BUILTIN_MESSAGE_UPSTREAM: &str = "c8y/s/us";
     const C8Y_TOPIC_ERROR_MESSAGE_DOWNSTREAM: &str = "c8y/s/e";
     const CLIENT_ID: &str = "check_connection_c8y";
-
-    let c8y_msg_pub_topic = Topic::new(C8Y_TOPIC_BUILTIN_MESSAGE_UPSTREAM)?;
-    let c8y_error_sub_topic = Topic::new(C8Y_TOPIC_ERROR_MESSAGE_DOWNSTREAM)?;
-
-    let mqtt = Client::connect(CLIENT_ID, &mqtt_client::Config::default().with_port(port)).await?;
-    let mut error_response = mqtt.subscribe(c8y_error_sub_topic.filter()).await?;
-
-    let (sender, mut receiver) = tokio::sync::oneshot::channel();
-
-    let _task_handle = tokio::spawn(async move {
-        while let Some(message) = error_response.next().await {
-            if message
-                .payload_str()
-                .unwrap_or("")
-                .contains("41,100,Device already existing")
-            {
-                let _ = sender.send(());
+    const REGISTRATION_PAYLOAD: &[u8] = b"100";
+    const REGISTRATION_ERROR: &[u8] = b"41,100,Device already existing";
+
+    let mut options = MqttOptions::new(CLIENT_ID, DEFAULT_HOST, port);
+    options.set_keep_alive(RESPONSE_TIMEOUT.as_secs() as u16);
+
+    let (mut client, mut connection) = rumqttc::Client::new(options, 10);
+    let mut acknowledged = false;
+
+    client.subscribe(C8Y_TOPIC_ERROR_MESSAGE_DOWNSTREAM, AtLeastOnce)?;
+
+    for event in connection.iter() {
+        match event {
+            Ok(Event::Incoming(Packet::SubAck(_))) => {
+                // We are ready to get the response, hence send the request
+                client.publish(
+                    C8Y_TOPIC_BUILTIN_MESSAGE_UPSTREAM,
+                    AtLeastOnce,
+                    false,
+                    REGISTRATION_PAYLOAD,
+                )?;
+            }
+            Ok(Event::Incoming(Packet::PubAck(_))) => {
+                // The request has been sent
+                acknowledged = true;
+            }
+            Ok(Event::Incoming(Packet::Publish(response))) => {
+                // We got a response
+                if &response.payload == REGISTRATION_ERROR {
+                    return Ok(DeviceStatus::AlreadyExists);
+                }
+            }
+            Ok(Event::Outgoing(Outgoing::PingReq)) => {
+                // No messages have been received for a while
+                println!("\nLocal MQTT publish has timed out.");
```

Comment:
I think we should remove `\n` here. Now output looks not so beautiful.

```
Sending packets to check connection. This may take up to 10 seconds.

Try 1 / 2: Sending a message to Cumulocity. 

Local MQTT publish has timed out.
... No response. If the device is new, it's normal to get no response in the first try.
Try 2 / 2: Sending a message to Cumulocity. 

Local MQTT publish has timed out.
... No response. 
Restarting mosquitto to resubscribe to bridged inbound cloud topics after device creation
Restarting mosquitto service.

Enabling software management.

```

Reply:
Fixed as suggested